mri/reconstructors/base.py

```diff
@@ -0,0 +1,146 @@
+                n_coils=n_coils)
+        elif fourier_type == 'stack':
+            self.fourier_op = Stacked3DNFFT(kspace_loc=kspace_loc,
+                                            shape=uniform_data_shape,
+                                            implementation=nfft_implementation,
+                                            n_coils=n_coils)
+        else:
+            raise ValueError('The value of fourier_type must be "cartesian" | '
+                             '"non-cartesian" | "stack"')
+        if verbose >= 5:
+            print("Initialized fourier operator : " + str(self.fourier_op))
+
+    def reconstruct(self, x_init=None):
+        """ This method calculates operator transform.
+        Parameters
+        ----------
+        x_init: np.ndarray (optional, default None)
+            input initial guess image for reconstruction
+        """
+        raise NotImplementedError("'reconstruct' is an abstract method.")
+
+
+class ReconstructorWaveletBase(ReconstructorBase):
+    """ This is the base reconstructor class for reconstruction.
+    This class holds some common parameters that is common for all
+    MR Image reconstructors
+    Parameters
+    ----------
+    kspace_loc: np.ndarray
+        the mask samples in the Fourier domain.
+    uniform_data_shape: tuplet
+        the shape of the matrix containing the uniform data.
+    n_coils: int, default 1
+        Number of coils used to acquire the signal in case of multiarray
+        receiver coils acquisition. If n_coils > 1, data shape must be
+        [n_coils, *data_shape]
+    fourier_type: str (optional, default 'cartesian')
+        type of fourier operator : 'cartesian' | 'non-cartesian' | 'stack'
+    nfft_implementation: str, default 'cpu'
+        way to implement NFFT : 'cpu' | 'cuda' | 'opencl'
+    wavelet_name: str | int
+        if implementation is with waveletN the wavelet name to be used during
+        the decomposition, else implementation with waveletUD2 where the
+        wavelet name is wavelet_id Refer to help of mr_transform under option
+        '-t' to choose the right wavelet_id.
+    padding_mode: str (optional, default zero)
+        The padding mode used in the Wavelet transform,
+        'zero' | 'periodization'
+    nb_scale: int (optional default is 4)
+        the number of scale in the used by the multi-scale wavelet
+        decomposition
+    wavelet_op_per_channel: bool
+        whether wavelet transform should be applied on every channel. This is
+        True only for Calibrationless reconstruction
+    n_jobs: int, default 1
+        Number of parallel jobs for linear operation
+    verbose: int
+        verbosity level for debug, please check dervied class for details
+        on verbosity levels
+    """
+    def __init__(self, kspace_loc, uniform_data_shape, n_coils,
+                 fourier_type, nfft_implementation, wavelet_name, padding_mode,
+                 nb_scale, wavelet_op_per_channel, n_jobs=1, verbose=0):
```

Comment:
I am sorry if I dont follow, same as ? 
I agree with you about the wavelet operator. It is just that this is very very redundant code if done in derived class. Do you have an alternative proposal?

Reply:
Something like:
```python
def __init__(self, wavelet_op, **kwargs):
	super(ReconstructorWaveletBase, self).__init__(**kwargs)
	self.linear_op = wavelet_op
```

But arguably, it's the same for `ReconstructorBase`, where you could just pass in directly a fourier operator. I juts have a hard time understanding why we want to instantiate the wavelet and the linear operator in these classes, rather than letting the user do it, giving them much more freedom. In particular, the amount of code to be written is going to be equivalent, so you are not winning in convention, but you do offer less flexibility.

Reply:
Wait, the goal is to prevent repeated codes right? 
If I move the operators to any other point within the package this would increase redundancy. 

Now coming to asking user input:
In my opinion, it is best to control on defining and using the operators on our own for:
1) The simplicity of use for user (I am not sure of the target users, but I am assuming they may not enough idea of the cost function being minimized)
2) Making our own scripts cleaner, with just a couple of parameters.
3) Inter-relations, this would help us to do some sanity checks before starting reconstruction,  which would surely save time if reconstruction takes too long.

I would like to hear @LElgueddari 's opinion on this too. Cause, I remember we came up with this coding convention for more reasons (ie moving away from operators defined from users)

Reply:
With offline discussion, we managed to see what could be abstracted out (gradients and so) and also how to reach a better tradeoff between convention and flexibility (having linear ops and fourier ops defined outside the reconstructor).

Reply:
@zaccharieramzi , I did some thinking and also had a discussion with @LElgueddari . 
I dont think we still dont agree with you based on the point 1) and 2) in my first reply. 

We genuinely feel that by giving all the freedom, we would eventually make the codes less user friendly. In the current approach, note that we have different default values for parameters, due to which, user can just choose basic parameters and in most cases just reconstruct directly with them.

Also, in our opinion, every reconstruct class implements a particular MR image model.

Coming to the point of reconstructing, by doing what you told, we would eventually be just thinning down the gap between the optimizers and reconstructors. If a user needs more flexibility, he is free to define the operators as he wants and directly call the fista / POGM / condatvu directly to obtain the image. 

I feel at this point we have reached with long github comments and many individual 1-1 discussions. I propose that @zaccharieramzi , @LElgueddari  and @chaithyagr sit down for a meeting and discuss and reach to a common ground on this and work towards it. 
Will set up a meeting this afternoon (if everyone is free). I will post the results of the discussion here. For now I will not implement the changes as described above.

Reply:
Summarizing meeting here:

We decided to go ahead and abstract out different operators in reconstructors class and have some defaults setup in case it has not been passed by the user.
This would ensure the data is abstracted, and ensures we can use **kwargs for any extra optimization parameters.

The end goal is to have reconstructors to take in linear, fourier, and proximity (or regularizer) as input and setup the gradient and call the right optimization. Later, the reconstructor could be used for reconstructing multiple images that were acquired in the same kspace location.

Reply:
Latest commit fixes this, resolving this here, please use new comments to start a new discussion.